In [1]:
import pandas as pd
import numpy as np

### Looking into the individual dataset of each month

### For January

In [2]:
january = pd.read_csv('jan.csv')
df_jan = pd.DataFrame(january)

In [3]:
df_jan.shape

(607346, 58)

In [4]:
df_jan.columns = df_jan.columns.str.lower()

In [5]:
df_jan['month'].unique()

array([1], dtype=int64)

In [6]:
df_jan.quarter.unique()

array([1], dtype=int64)

In [7]:
df_jan.columns

Index(['year', 'quarter', 'month', 'day_of_month', 'day_of_week', 'fl_date',
       'op_unique_carrier', 'op_carrier_airline_id', 'op_carrier', 'tail_num',
       'op_carrier_fl_num', 'origin_airport_id', 'origin_airport_seq_id',
       'origin_city_market_id', 'origin', 'origin_city_name',
       'origin_state_abr', 'origin_state_fips', 'origin_state_nm',
       'origin_wac', 'dest_airport_id', 'dest_airport_seq_id',
       'dest_city_market_id', 'dest', 'dest_city_name', 'dest_state_abr',
       'dest_state_fips', 'dest_state_nm', 'dest_wac', 'dep_time', 'dep_delay',
       'dep_delay_new', 'dep_time_blk', 'taxi_out', 'wheels_off', 'wheels_on',
       'taxi_in', 'arr_time', 'arr_delay', 'arr_delay_new', 'cancelled',
       'diverted', 'crs_elapsed_time', 'actual_elapsed_time', 'air_time',
       'flights', 'distance', 'carrier_delay', 'weather_delay', 'nas_delay',
       'security_delay', 'late_aircraft_delay', 'div_airport_landings',
       'div_reached_dest', 'div_actual_elapsed_ti

### For February

In [8]:
feb = pd.read_csv('feb.csv')
df_feb = pd.DataFrame(feb)

In [9]:
df_feb.shape

(574268, 58)

In [10]:
df_feb.columns

Index(['YEAR', 'QUARTER', 'MONTH', 'DAY_OF_MONTH', 'DAY_OF_WEEK', 'FL_DATE',
       'OP_UNIQUE_CARRIER', 'OP_CARRIER_AIRLINE_ID', 'OP_CARRIER', 'TAIL_NUM',
       'OP_CARRIER_FL_NUM', 'ORIGIN_AIRPORT_ID', 'ORIGIN_AIRPORT_SEQ_ID',
       'ORIGIN_CITY_MARKET_ID', 'ORIGIN', 'ORIGIN_CITY_NAME',
       'ORIGIN_STATE_ABR', 'ORIGIN_STATE_FIPS', 'ORIGIN_STATE_NM',
       'ORIGIN_WAC', 'DEST_AIRPORT_ID', 'DEST_AIRPORT_SEQ_ID',
       'DEST_CITY_MARKET_ID', 'DEST', 'DEST_CITY_NAME', 'DEST_STATE_ABR',
       'DEST_STATE_FIPS', 'DEST_STATE_NM', 'DEST_WAC', 'DEP_TIME', 'DEP_DELAY',
       'DEP_DELAY_NEW', 'DEP_TIME_BLK', 'TAXI_OUT', 'WHEELS_OFF', 'WHEELS_ON',
       'TAXI_IN', 'ARR_TIME', 'ARR_DELAY', 'ARR_DELAY_NEW', 'CANCELLED',
       'DIVERTED', 'CRS_ELAPSED_TIME', 'ACTUAL_ELAPSED_TIME', 'AIR_TIME',
       'FLIGHTS', 'DISTANCE', 'CARRIER_DELAY', 'WEATHER_DELAY', 'NAS_DELAY',
       'SECURITY_DELAY', 'LATE_AIRCRAFT_DELAY', 'DIV_AIRPORT_LANDINGS',
       'DIV_REACHED_DEST', 'DIV_ACTUAL_ELAPSED_TI

In [11]:
df_feb.columns = df_feb.columns.str.lower()

In [12]:
df_feb.month.unique()

array([2], dtype=int64)

In [13]:
df_feb.quarter.unique()

array([1], dtype=int64)

### For March

In [14]:
mar = pd.read_csv('mar.csv')
df_mar = pd.DataFrame(mar)

In [15]:
df_mar.shape

(648229, 58)

In [16]:
df_mar.columns = df_mar.columns.str.lower()

In [17]:
df_mar.month.unique()

array([3], dtype=int64)

In [18]:
df_mar.quarter.unique()

array([1], dtype=int64)

### For April

In [19]:
apr = pd.read_csv('apr.csv')
df_apr = pd.DataFrame(apr)

In [20]:
df_apr.shape

(313382, 58)

In [21]:
df_apr.columns = df_apr.columns.str.lower()

In [22]:
df_apr.month.unique()

array([4], dtype=int64)

In [23]:
df_apr.quarter.unique()

array([2], dtype=int64)

### For May

In [24]:
may = pd.read_csv('may.csv')
df_may = pd.DataFrame(may)

In [25]:
df_may.shape

(180617, 58)

In [26]:
df_may.columns = df_may.columns.str.lower()

In [27]:
df_may.month.unique()

array([5], dtype=int64)

In [28]:
df_may.quarter.unique()

array([2], dtype=int64)

### For June

In [29]:
jun = pd.read_csv('jun.csv')
df_jun = pd.DataFrame(jun)

In [30]:
df_jun.shape

(223732, 58)

In [31]:
df_jun.columns = df_jun.columns.str.lower()

In [32]:
df_jun.month.unique()

array([6], dtype=int64)

In [33]:
df_jun.quarter.unique()

array([2], dtype=int64)

## Bringing all the datasets together

In [34]:
df = pd.concat([df_jan, df_feb, df_mar, df_apr, df_may, df_jun])
df.shape

(2547574, 58)

<b>Bringing the datasets together gave a sum of <i>2,547,574</i> rows.</b>

<b>Q1: What airline had the most number of flights in the first quarter of 2019 ?</b>

In [35]:
df.columns

Index(['year', 'quarter', 'month', 'day_of_month', 'day_of_week', 'fl_date',
       'op_unique_carrier', 'op_carrier_airline_id', 'op_carrier', 'tail_num',
       'op_carrier_fl_num', 'origin_airport_id', 'origin_airport_seq_id',
       'origin_city_market_id', 'origin', 'origin_city_name',
       'origin_state_abr', 'origin_state_fips', 'origin_state_nm',
       'origin_wac', 'dest_airport_id', 'dest_airport_seq_id',
       'dest_city_market_id', 'dest', 'dest_city_name', 'dest_state_abr',
       'dest_state_fips', 'dest_state_nm', 'dest_wac', 'dep_time', 'dep_delay',
       'dep_delay_new', 'dep_time_blk', 'taxi_out', 'wheels_off', 'wheels_on',
       'taxi_in', 'arr_time', 'arr_delay', 'arr_delay_new', 'cancelled',
       'diverted', 'crs_elapsed_time', 'actual_elapsed_time', 'air_time',
       'flights', 'distance', 'carrier_delay', 'weather_delay', 'nas_delay',
       'security_delay', 'late_aircraft_delay', 'div_airport_landings',
       'div_reached_dest', 'div_actual_elapsed_ti

In [36]:
df.drop(columns=['unnamed: 57'], inplace=True)

In [37]:
df['op_unique_carrier'].nunique() == df['op_carrier'].nunique() == df['op_carrier_airline_id'].nunique()

True

From the preceding output, i see that the <b>op_unique_carrier, op_carrier and the, op_carrier_airline_id</b> columns all have the same number of airlines which is 17.

In [38]:
#Using the op_unique_carrier column, i'll answer the question
df[df['quarter']==1].groupby('op_unique_carrier')['flights'].count().nlargest(1)

op_unique_carrier
WN    328040
Name: flights, dtype: int64

<b>The WN airline(carrier) has the most amount of flights in the first Quarter of 2019.</b>

In [39]:
df[df['quarter']==1]['month'].unique()

array([1, 2, 3], dtype=int64)

### What is the total number of airline travels in the first half of 2019?

In [40]:
df['flights'].isnull().sum()

0

In [41]:
df.shape[0]

2547574

<b>There was a total of 2,547,574 flights throughout the first half of 2019.</b>

### What is the total airtime by all airlines in may 2019?

In [42]:
df['air_time'] # Airtime is in minutes.

0         74.0
1         71.0
2         57.0
3         63.0
4         57.0
          ... 
223727    41.0
223728    37.0
223729    36.0
223730    36.0
223731    41.0
Name: air_time, Length: 2547574, dtype: float64

In [43]:
df['air_time'].nlargest(1) 

26698    698.0
Name: air_time, dtype: float64

In [44]:
df[df['month']==5]['air_time'].sum()

16909348.0

In [45]:
#Converting our time in minutes to hours
df[df['month']==5]['air_time'].sum()/60

281822.4666666667

#### There was approximately 281,822 hours of flight throughout may  2019.

### What city do you consider as the busiest in april 2019?

This question considers factors such as air_time or number of flights. For this analysis, i'll use the number of flights because air_time differs;  it could be long or short while number of flights records the amount of flights. Also, the origin_city_name column will be used since that's where the flight takes off from.

In [46]:
df[df['month']==4].groupby('origin_city_name')['flights'].sum().nlargest(1)

origin_city_name
Chicago, IL    18997.0
Name: flights, dtype: float64

#### From my analysis, CHICAGO has the most amount of flights in april 2019, which makes it considerably the busiest.

### What airline covered the most distance in june 2019?

In [47]:
df[df['month']==6].groupby('op_unique_carrier')['distance'].sum().nlargest(1) # NOTE: The distance is in Miles

op_unique_carrier
WN    47424693.0
Name: distance, dtype: float64

From the analysis, WN airline(carrier) covered the most distance in june 2019.

### What is the tail number of the aircraft that covered the most distance in the 2nd quarter of 2019?

In [48]:
df[df['quarter']==2].groupby('tail_num')['distance'].sum().nlargest(1)

tail_num
N820DX    424613.0
Name: distance, dtype: float64

From the preceding output, the aircraft with the tail number N820DX covered the most distance of 424,613 miles.

### Which state recoreded the most delays in  the 2nd quarter of 2019?

According to the data given, there are two delays(in minutes) recorded; arrival delay and departure delay(arr_delay_new, dep_delay_new). To calculate the most delays, we'll add up both delays.

In [57]:
df['total_delay'] = df['arr_delay_new'] + df['dep_delay_new']

In [58]:
df['total_delay'] # total delay i.e arrival and departure delay summed up

0         109.0
1           0.0
2           0.0
3          20.0
4           1.0
          ...  
223727      2.0
223728      9.0
223729      2.0
223730     15.0
223731      0.0
Name: total_delay, Length: 2547574, dtype: float64

In [66]:
df[df['quarter']==2].groupby('origin_state_nm')['total_delay'].sum().nlargest(1)/60

origin_state_nm
Texas    10433.333333
Name: total_delay, dtype: float64

Texas has the most delays in total(arrival and departure delays.)

### What is the daily average taxi-out time in New York in February 2019?

In [80]:
df[(df['origin_state_nm']=='New York') & (df['month']==2)]['taxi_out']\
                                         .sum()/(len(df[df['origin_state_nm']=='New York']['taxi_out']))

5.507048901475683

### How  many airlines with New York destinations were diverted in Q1 2019?

In [86]:
df_st_qu = df[(df['dest_state_nm']=='New York') & (df['quarter']==1)]
df_st_qu.groupby('op_unique_carrier')['div_airport_landings'].count()

op_unique_carrier
9E    13182
AA    10765
AS     1127
B6    14563
DL    17456
EV      946
F9     1272
G4      936
HA       89
MQ     3512
NK     1167
OH     2083
OO     4918
UA     1917
WN     6718
YV      623
YX    11541
Name: div_airport_landings, dtype: int64

The entire 17 airlines with New York destination were diverted.

### Which of the delays is more rampant in washington in the first half of 2019?

In [91]:
df[df['origin_state_nm']=='Washington']['weather_delay'].sum()

18816.0

In [92]:
df[df['origin_state_nm']=='Washington']['security_delay'].sum()

686.0

In [93]:
df[df['origin_state_nm']=='Washington']['late_aircraft_delay'].sum()

80962.0

In [95]:
df[df['origin_state_nm']=='Washington']['nas_delay'].sum()

74437.0

In [96]:
df[df['origin_state_nm']=='Washington']['carrier_delay'].sum()

159591.0

The carrier delay is the most rampant in Washington.